src/array.cr

```diff
@@ -1521,7 +1521,7 @@ class Array(T)
     result
   end
 
-  def product(ary, &block)
+  def product(ary : Array(U), &block)
```

Comment:
Here it's better if we do `ary : Enumerable` because only `each` is needed from `ary` (the argument should probably be named `enumerable`, but it's also OK if we leave it like `ary`).